crates/bevy_ecs/src/query/state.rs

```diff
@@ -341,16 +341,55 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
     /// If `world` does not match the one used to call `QueryState::new` for this instance.
     pub fn update_archetypes_unsafe_world_cell(&mut self, world: UnsafeWorldCell) {
         self.validate_world(world.id());
-        let archetypes = world.archetypes();
-        let old_generation =
-            std::mem::replace(&mut self.archetype_generation, archetypes.generation());
-
-        for archetype in &archetypes[old_generation..] {
-            // SAFETY: The validate_world call ensures that the world is the same the QueryState
-            // was initialized from.
-            unsafe {
-                self.new_archetype_internal(archetype);
+        if self.component_access.required.is_empty() {
+            let archetypes = world.archetypes();
+            let old_generation =
+                std::mem::replace(&mut self.archetype_generation, archetypes.generation());
+
+            for archetype in &archetypes[old_generation..] {
+                // SAFETY: The validate_world call ensures that the world is the same the QueryState
+                // was initialized from.
+                unsafe {
+                    self.new_archetype_internal(archetype);
+                }
             }
+        } else {
+            // skip if we are already up to date
+            if self.archetype_generation == world.archetypes().generation() {
```

Comment:
I don't know if this is faster to be doing in the case we are updating an already partially populated cache although, as always, it depends heavily on usage patterns. Looking through all the required components every time seems like a lot of jumping around in memory. Given that archetype deletion doesn't yet exist it seems like it the way to do this efficiently would be to use the same component every time and keep track of which archetype id you were up to, but that requires tracking additional state.

All of this is intended to be replace with observers in short order anyway so I don't think this implementation is a deal breaker in the meantime. Big improvement for populating the cache of a new query in a world with a large number of archetypes.